[Audit logs] Process events

remidej opened the pull request:
### What does it do?

Creates a new processEvent function inside the audit logs service, with the following responsibilities:

- Filtering events: only those in the map are handled, the others ignored
- Building the payload object based on the arguments received by the event hub
- Sanitizing the payload
- Building an object ready to be passed to the storage providers

### What does it NOT do:

- Create a provider system for the filtering. Our current idea is that the eventMap object could be extracted into a provider system in the future.
- Sanitize all the payloads. It doesn't do it for CRUD events, only for auth

### How to test it?

Do actions in Strapi, see what gets logged, see what gets ignored

co-authors @markkaylor @Feranchz

markkaylor commented on packages/core/admin/ee/server/services/audit-logs/process-event.js:
```diff
@@ -0,0 +1,33 @@
+'use strict';
+
+module.exports = function processEvent(strapi, name, ...args) {
+  const defaultGetPayload = (...args) => args[0];
```
Maybe `getDefaultPayload`

remidej replied:
sure

markkaylor commented on packages/core/admin/ee/server/services/audit-logs/process-event.js:
```diff
@@ -0,0 +1,33 @@
+module.exports = function processEvent(strapi, name, ...args) {
+  const defaultGetPayload = (...args) => args[0];
+  const getAuthPayload = (...args) => ({
+    user: strapi.service('admin::user').sanitizeUser(args[0].user),
```
I see the sanitization is done on emit for the entity manager:
https://github.com/strapi/strapi/blob/972bdd72cc07f8ae378f472162cd0f4371aa3f26/packages/core/content-manager/server/services/entity-manager.js#L15  Should we be doing it here or when we emit events? If we do it here should we refactor the entity manager `wrapEmitEvent`?

remidej replied:
Very interesting, thanks for linking this. I think it's probably better to sanitize when you're emitting the event. You probably don't want sensitive data in your event hub at all, it's not just about audit logs.

So I'd vote for not changing the event manager, removing our process event sanitizing, and making sure we sanitize the data in other places where events are emitted

markkaylor commented on packages/core/admin/ee/server/services/audit-logs/process-event.js:
```diff
@@ -0,0 +1,33 @@
+'use strict';
+
+module.exports = function processEvent(strapi, name, ...args) {
```
It would be nice to know if we need to spread `...args` or not. I get the feeling this parameter could just be called `payload`. Then we wouldn't need the `defaultGetPayload` function

remidej replied:
Currently it's not used, but we allow passing as many args as wanted in the event hub, so if we don't spread here we lose access to everything but the 1st arg. I think if we wanted to prevent passing several args we should have done it in the event emitter, but we wanted to follow the node event emitter behavior there

remidej replied:
The 1 place where we could easily remove args[0] is the default get payload:

```js
const getDefaultPayload = payload => payload
```

looks a bit weird but it will work

remidej replied:
If we want to keep as simple as it can be that also works for me. But in that case I think we can remove the object altogether and just keep an array of allowed events, since there's no point in having 1 function per event anymore

markkaylor replied:
> I think if we wanted to prevent passing several args we should have done it in the event emitter, but we wanted to follow the node event emitter behavior there 

The eventhub is used elsewhere in the application and is not specific to audit logs, so I think it still makes sense to accept many arguments there. It seems to me (at least so far) we will always just pass a payload for audit logs.

I guess though I'm having a bit of a tough time understanding when we would need additional arguments here for audit logs.

So far though I think the map does look like a bit of overkill because they all accept the same `getDefaultPayload` and that function itself doesn't do much. @remidej maybe you could provide an example for what you see in the future?

I do think providers should be able to extend this map with their own function, but not sure that is dependant on the `...args` argument. 

I brought this up because still wasn't clear to me why we needed it and I'll be off for a while. But either way we can always change it later even after audit logs is released, I don't think it would be a breaking changing. So we can keep moving forward either way in my opinion.

markkaylor commented on packages/core/admin/ee/server/services/audit-logs/process-event.js:
```diff
@@ -0,0 +1,33 @@
+  });
+
+  const eventMap = {
+    'entry.create': defaultGetPayload,
```
What's the advantage of storing a function reference here vs invoking the function and storing its value. Especially if we can pass the payload directly instead of `...args`.

remidej replied:
I think to call the function directly you'd need several if/else statements or a switch case. I find the object a bit more concise. Also I think it's easier to extend an object via a provider system

remidej replied:
But as I [commented](https://github.com/strapi/strapi/pull/15179#discussion_r1050616327) above, if we don't care about having a custom payload for an event, then we don't need the object

remidej commented:
@Feranchz @markkaylor based on the conversations above I tried something a little different. We agreed that it was weird to have to repeat getDefaultPayload for all default events. So instead I extracted a getEventMap function that receives the array of default events to build that object.

I think that `getEventMap` function can be good for composition as well, because it could be replaced by providers or user configs in the future. They have access to the list of default events, so they can remove some:

```js
const getEventMap = (defaultEvents) => {
  const getDefaultPayload = (...args) => args[0];

  const blacklist = ['media.create', 'media.update', 'media.delete'];

  // Use the default payload for all default events
  return defaultEvents.reduce((acc, event) => {
    if (!blacklist.includes(event)) {
      acc[event] = getDefaultPayload;
    }
    return acc;
  }, {});
};
```

or add some, while also keeping the ability to handle events that receive many params:

```js
const getEventMap = (defaultEvents) => {
  const getDefaultPayload = (...args) => args[0];

  // Use the default payload for all default events
  const defaultMap = defaultEvents.reduce((acc, event) => {
    acc[event] = getDefaultPayload;
    return acc;
  }, {});

  return {
    ...defaultMap,
    'entry.read': getDefaultPayload,
    'sentry-plugin.emit': (error, scope) => ({ error, scope }),
  }
};
```

I also realized that it was not necessary to build the event map object every time an event is emitted, so I moved that logic to when the service is created. This also meant there was very little left in the process events file, so I moved the code back to the main service file.
